pkg/performanceprofile/profilecreator/profilecreator.go

```diff
@@ -289,30 +289,108 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
 	return disabledHTTopology
 }
 
-// GetReservedAndIsolatedCPUs returns Reserved and Isolated CPUs
-func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, splitReservedCPUsAcrossNUMA bool, disableHTFlag bool) (cpuset.CPUSet, cpuset.CPUSet, error) {
+// Calculates the resevered, isolated and offlined cpuSets.
+func (ghwHandler GHWHandler) CalculateCPUSets(reservedCPUCount int, offlinedCPUCount int, splitReservedCPUsAcrossNUMA bool, disableHTFlag bool) (cpuset.CPUSet, cpuset.CPUSet, cpuset.CPUSet, error) {
+
 	cpuInfo, err := ghwHandler.CPU()
 	if err != nil {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't obtain CPU info from GHW snapshot: %v", err)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't obtain CPU info from GHW snapshot: %v", err)
 	}
 
+	// Check limits are in range
 	if reservedCPUCount <= 0 || reservedCPUCount >= int(cpuInfo.TotalThreads) {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
+	}
+
+	if offlinedCPUCount >= int(cpuInfo.TotalThreads) {
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the offlined CPU count in the range [0,%d]", cpuInfo.TotalThreads-1)
 	}
+
+	if reservedCPUCount+offlinedCPUCount >= int(cpuInfo.TotalThreads) {
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved plus offlined CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
+	}
+
+	// Get system info
 	topologyInfo, err := ghwHandler.SortedTopology()
 	if err != nil {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't obtain Topology Info from GHW snapshot: %v", err)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't obtain Topology Info from GHW snapshot: %v", err)
 	}
+
 	htEnabled, err := ghwHandler.IsHyperthreadingEnabled()
 	if err != nil {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't determine if Hyperthreading is enabled or not: %v", err)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't determine if Hyperthreading is enabled or not: %v", err)
+	}
+
+	//Calculate offlined cpus
+	offlined, err := ghwHandler.GetOfflinedCPUs(cpuInfo, topologyInfo, offlinedCPUCount, disableHTFlag, htEnabled)
```

Comment:
so offlined takes precedence over reserved and isolated when computing cpu set partitioning, right?
Would it be simpler (and safer) if we compute first the reserved set, then the offlined set and then we trivially (it seems trivial at least) compute the isolated set as the remaining CPU set?

Reply:
I have nothing against doing it as you propose but I do not understand why it would be simpler and safer to calculate first reserved. 

I thought of calculating offlined first because:
- if user do not want to disable HT and offlinedCount is enough to disable a complete socket we can do it to save some power ( as far as I understood @marioferh  measurements that's the only scenario where some power saving occurs)
- otherwise I will try to offline the sibling threads to help with performance.

Reply:
the reasoning is that reserved is compulsory, and the reserved set is the most important set because it's needed for the node to be functional. IOW, without proper setting reserved set, we can kill the node, which is basically among the worst possible outcomes.

Reply:
ok, let me then change the order in which they are calculated.